### ISSUE TITLE
clj-kondo.exe native-image binary fails to work with --lint - (stdin) in PowerShell

### ISSUE THREAD

**Issue by sogaiu** [borkdude/clj-kondo#309]
**version**

ecc93060

**platform**

Windows 10

**problem**

clj-kondo.exe --lint - (stdin) does not work in PowerShell (Windows and Core)

**repro**

Start PowerShell (Windows or Core), and then:

```
cd c:\
git clone https://github.com/clojure/clojure
type c:\clojure\src\clj\clojure\core.clj | clj-kondo.exe --lint -
REM press Enter more than once if necessary
```

**expected behavior**

linted output

**actual behavior**

Output seen:

```
-:0:0: error: could not process file
linting took 17310ms, errors: 1, warnings: 0
```

**Comment by borkdude** [comment]
@sogaiu Have you tried it like in this SO answer? https://stackoverflow.com/a/11788475/6264

**Comment by sogaiu** [comment]
As discussed elsewhere, it appears native-image on Windows doesn't work so well with Clojure's \*in\* and \*out\*.  I don't know why.

The following is a patch that avoids the use of \*in\* and binds \*out\* to System/out.  It seems to work here.

```
diff --git a/src/clj_kondo/impl/core.clj b/src/clj_kondo/impl/core.clj
index 34ee95d..bc6442e 100644
--- a/src/clj_kondo/impl/core.clj
+++ b/src/clj_kondo/impl/core.clj
@@ -142,7 +142,7 @@
                                    dev?)
                (sources-from-dir file)))
         (= "-" filename)
-        [(ana/analyze-input ctx "<stdin>" (slurp *in*) default-language dev?)]
+        [(ana/analyze-input ctx "<stdin>" (slurp System/in) default-language dev?)]
         (classpath? filename)
         (mapcat #(process-file ctx % default-language)
                 (str/split filename
diff --git a/src/clj_kondo/main.clj b/src/clj_kondo/main.clj
index 8d2c11b..39f25c1 100644
--- a/src/clj_kondo/main.clj
+++ b/src/clj_kondo/main.clj
@@ -81,10 +81,11 @@ Options:
                  :else (let [{:keys [:summary]
                               :as results} (clj-kondo/run! parsed)
                              {:keys [:error :warning]} summary]
-                         (clj-kondo/print! results)
-                         (cond (pos? error) 3
-                               (pos? warning) 2
-                               :else 0)))
+                         (binding [*out* (java.io.OutputStreamWriter. System/out)]
+                           (clj-kondo/print! results)
+                           (cond (pos? error) 3
+                                 (pos? warning) 2
+                                 :else 0))))
            0)))
     (finally
       (profiler/print-profile :main))))
```

**Comment by sogaiu** [comment]
> @sogaiu Have you tried it like in this SO answer? https://stackoverflow.com/a/11788475/6264

The result appears the same here.

**Comment by borkdude** [comment]
@sogaiu This is what I'm seeing with a fresh install of Powershell on a MBP:

```
PS /Users/borkdude/Dropbox/dev/clojure/clj-kondo> Get-Content ./src/clj_kondo/main.clj | clj-kondo --lint -
linting took 23ms, errors: 0, warnings: 0
```

**Comment by sogaiu** [comment]
When you say MBP, does that mean you are using macos?

**Comment by borkdude** [comment]
@sogaiu Correct. You mentioned Core so I assumed this affected other operating systems as well.

**Comment by sogaiu** [comment]
My suspicion is that this is a native-image on Windows issue.

**Comment by sogaiu** [comment]
Possibly some more clues...

On windows 10 with native-image:

```
(ns script
  (:gen-class))

(defn -main [& args]
  (println (slurp *in*)))
```

yields a stacktrace involving jni and java.io.FileInputStream.readBytes:

```
Exception in thread "System-2" java.io.IOException: The handle is invalid
        at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VA_LIST:Ljava_io_IOException_2_0002e_0003cinit_0003e_00028Ljava_lang_String_2_00029V(JNIJavaCallWrappers.java:0)
        at java.io.FileInputStream.readBytes(FileInputStream.java)
        at java.io.FileInputStream.read(FileInputStream.java:255)
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:284)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
        at java.io.InputStreamReader.read(InputStreamReader.java:184)
        at java.io.BufferedReader.read1(BufferedReader.java:210)
        at java.io.BufferedReader.read(BufferedReader.java:286)
        at java.io.LineNumberReader.read(LineNumberReader.java:166)
        at java.io.FilterReader.read(FilterReader.java:74)
        at java.io.PushbackReader.read(PushbackReader.java:128)
        at java.io.BufferedReader.fill(BufferedReader.java:161)
        at java.io.BufferedReader.read1(BufferedReader.java:212)
        at java.io.BufferedReader.read(BufferedReader.java:286)
        at java.io.Reader.read(Reader.java:140)
        at clojure.java.io$fn__11010.invokeStatic(io.clj:337)
        at clojure.java.io$fn__11010.invoke(io.clj:334)
        at clojure.lang.MultiFn.invoke(MultiFn.java:238)
        at clojure.java.io$copy.invokeStatic(io.clj:406)
        at clojure.java.io$copy.doInvoke(io.clj:391)
        at clojure.lang.RestFn.invoke(RestFn.java:425)
        at clojure.core$slurp.invokeStatic(core.clj:6871)
        at clojure.core$slurp.doInvoke(core.clj:6862)
        at clojure.lang.RestFn.invoke(RestFn.java:410)
        at script$_main.invokeStatic(script.clj:5)
        at script$_main.doInvoke(script.clj:4)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at script.main(Unknown Source)
```

IIUC, readBytes (see close to the top of the stacktrace) ends up being a native call -- I think that means that platform specific-bits could be involved.

Also, according to:

https://github.com/clojure/clojure/blob/653b8465845a78ef7543e0a250078eea2d56b659/src/jvm/clojure/lang/RT.java#L185-L187

*in* is:

```
final static public Var IN =
		Var.intern(CLOJURE_NS, Symbol.intern("*in*"),
		           new LineNumberingPushbackReader(new InputStreamReader(System.in))).setDynamic();

```

The following doesn't yield a stacktrace via native-image on Windows and it does produce the expected output:

```
(ns script
  (:gen-class))

(defn -main [& args]
  (println (slurp (clojure.lang.LineNumberingPushbackReader. (java.io.InputStreamReader. System/in)))))
```

The testing is from powershell and cmd.exe, with something like:

```
type project.clj | .\my-native-image.exe
```

I believe the original snippet involving *in* works on various linux distributions and macos, though I haven't recently tested the latter.

In summary, the two differences appear to be:

* Windows exhibits the issue, but the other platforms don't
* setDynamic() seems to make a difference

**Comment by borkdude** [comment]
Fixed with GraalVM 19.3.0
